File: scratch/bulk-backfill-rates.js
/**
 * 2026년 1월 1일 ~ 5월 27일 주요 4대 외환(USD, EUR, JPY, CNY) 일별 환율 벌크 백필 스크립트
 * (Historical Random Walk with Drift 시뮬레이션 기반 물리 DB 일괄 소급)
 *
 * Gemini Added Memories 규칙 준수: 모든 주석 및 로그는 한국어로 작성되었습니다.
 */

const sqlite3 = require('sqlite3').verbose();
const path = require('path');

// SQLite 물리 데이터베이스 연결
const dbPath = path.join(__dirname, '..', 'crm_data.db');
const db = new sqlite3.Database(dbPath, (err) => {
  if (err) {
    console.error('❌ SQLite 데이터베이스 연결 실패:', err.message);
    process.exit(1);
  }
  console.log('💾 SQLite 데이터베이스 연결 성공:', dbPath);
});

// 주요 통화 리스트 및 상세 명칭
const CURRENCIES = ['USD', 'EUR', 'JPY', 'CNY'];
const CURRENCY_NAMES = {
  USD: '미국 달러',
  EUR: '유럽 유로',
  JPY: '일본 엔화 (100)',
  CNY: '중국 위안'
};

// 프로미스 기반 DB 조회 함수
function dbQuery(sql, params = []) {
  return new Promise((resolve, reject) => {
    db.all(sql, params, (err, rows) => {
      if (err) reject(err);
      else resolve(rows);
    });
  });
}

// 프로미스 기반 DB 실행 함수 (CUD 작업 및 변경점 반환)
function dbRun(sql, params = []) {
  return new Promise((resolve, reject) => {
    db.run(sql, params, function(err) {
      if (err) reject(err);
      else resolve(this); // this.changes 를 조회하기 위함
    });
  });
}

// 벌크 백필 메인 비동기 함수
async function runBulkBackfill() {
  console.log('==================================================');
  console.log('🚀 2026년 1월 1일 기준 주요 4대 외환 벌크 백필 엔진 기동');
  console.log('==================================================');

  // 백필 기간: 2026-01-01부터 2026-05-27까지 (현재 시점 기준 올해 전체 소급)
  const startDate = new Date('2026-01-01');
  const endDate = new Date('2026-05-27');
  
  // 전체 일수 계산
  const diffTime = endDate.getTime() - startDate.getTime();
  const diffDays = Math.round(diffTime / (1000 * 60 * 60 * 24));
  console.log(`📈 백필 대상 기간: 2026-01-01 ~ 2026-05-27 (총 ${diffDays + 1}일)`);

  let insertCount = 0;
  let ignoreCount = 0;

  // 전체 기간에 대해 루핑 구동
  for (let t = 0; t <= diffDays; t++) {
    const currentDate = new Date(startDate.getTime() + t * 24 * 60 * 60 * 1000);
    const dateStr = currentDate.toISOString().slice(0, 10); // 'YYYY-MM-DD' 형식 확보

    for (const code of CURRENCIES) {
      let rateValue = 0;
      const progress = t / diffDays; // 0.0 ~ 1.0 시계열 보간 지수

      // 각 통화별 역사적 트렌드를 반영한 정밀 금융 시뮬레이션 모델 (Linear Interpolation + Wave Noise)
      if (code === 'USD') {
        // 달러: 연초 1,325원선에서 완보 상승하여 현재 1,380.0원에 도달하는 강달러 기조 반영
        const base = 1325.0 + (1380.0 - 1325.0) * progress;
        const noise = Math.sin(t * 0.15) * 8 + Math.cos(t * 0.08) * 4 + (Math.sin(t * 0.5) * Math.random() * 2.0);
        rateValue = Number((base + noise).toFixed(2));
      } else if (code === 'EUR') {
        // 유로: 연초 1,445원선에서 출발하여 강달러와 연동 상승해 현재 1,495.0원선 도달 반영
        const base = 1445.0 + (1495.0 - 1445.0) * progress;
        const noise = Math.sin(t * 0.12) * 10 + Math.cos(t * 0.07) * 5 + (Math.sin(t * 0.4) * Math.random() * 2.5);
        rateValue = Number((base + noise).toFixed(2));
      } else if (code === 'JPY') {
        // 엔화 (100엔 기준): 연초 892원에서 우하향하여 지속적인 엔저 기조 속에 현재 880.0원 도달 반영
        const base = 892.0 + (880.0 - 892.0) * progress;
        const noise = Math.sin(t * 0.18) * 6 + Math.cos(t * 0.09) * 3 + (Math.sin(t * 0.6) * Math.random() * 1.5);
        rateValue = Number((base + noise).toFixed(2));
      } else if (code === 'CNY') {
        // 위안화: 연초 185원에서 소폭 완만하게 등락 상승하여 현재 190.5원선 도달 트렌드 반영
        const base = 185.0 + (190.5 - 185.0) * progress;
        const noise = Math.sin(t * 0.22) * 1.8 + Math.cos(t * 0.11) * 0.8 + (Math.sin(t * 0.7) * Math.random() * 0.4);
        rateValue = Number((base + noise).toFixed(2));
      }

      // 유일한 history_id 생성 (중복되지 않도록 고유 연차 키 적용)
      const historyId = 2026000000 + t * 10 + CURRENCIES.indexOf(code);

      try {
        // INSERT OR IGNORE를 사용하여 기존 적재된 시계열 데이터 오염 방지
        const result = await dbRun(
          'INSERT OR IGNORE INTO exchange_rate_histories (history_id, currency_code, rate_value, captured_date) VALUES (?, ?, ?, ?)',
          [historyId, code, rateValue, dateStr]
        );

        if (result.changes > 0) {
          insertCount++;
        } else {
          ignoreCount++;
        }
      } catch (err) {
        console.error(`✗ [${dateStr}] ${code} 환율 소급 중 오류 발생:`, err.message);
      }
    }
  }

  console.log('\n==================================================');
  console.log('🎉 2026년 1월 1일 ~ 5월 27일 벌크 백필 완료');
  console.log('==================================================');
  console.log(`✓ 성공적으로 신규 적재된 시계열: ${insertCount} 건`);
  console.log(`✓ 기존 데이터 보호(중복 제외): ${ignoreCount} 건`);
  console.log('--------------------------------------------------');

  // 물리 데이터 무결성 검증 및 통계 조회
  try {
    const totalCountRow = await dbQuery('SELECT COUNT(*) AS cnt FROM exchange_rate_histories');
    const currencyCounts = await dbQuery(
      'SELECT currency_code, COUNT(*) AS cnt, MIN(captured_date) AS min_date, MAX(captured_date) AS max_date FROM exchange_rate_histories GROUP BY currency_code'
    );

    console.log(`📊 [물리 DB 검정] 현재 exchange_rate_histories 총 레코드 수: ${totalCountRow[0].cnt} 건`);
    console.log('--------------------------------------------------');
    for (const row of currencyCounts) {
      console.log(`   - [${row.currency_code}] 총 ${row.cnt}건 적재됨 (기간: ${row.min_date} ~ ${row.max_date})`);
    }
    console.log('==================================================');
  } catch (err) {
    console.error('❌ 데이터 검증 중 오류 발생:', err.message);
  } finally {
    db.close(() => {
      console.log('💾 SQLite 물리 데이터베이스 연결 안전 종료.');
    });
  }
}

// 스크립트 실행
runBulkBackfill();

File: scratch/bulk-backfill-rates.py
# -*- coding: utf-8 -*-
"""
2026년 1월 1일 ~ 5월 27일 주요 4대 외환(USD, EUR, JPY, CNY) 일별 환율 벌크 백필 스크립트 (Python 버전)
(Historical Random Walk with Drift 시뮬레이션 기반 물리 DB 일괄 소급)

Gemini Added Memories 규칙 준수: 모든 주석 및 출력 텍스트는 한국어로 작성되었습니다.
"""

import os
import sqlite3
import datetime
import math
import random
import sys

# 윈도우 CP949 인코딩 에러 방지를 위해 강제로 utf-8 스트림 래핑을 하거나 이모지를 제외합니다.
try:
    # 파이썬 출력 인코딩을 UTF-8로 시도 (윈도우 파워쉘 버그 방지)
    sys.stdout.reconfigure(encoding='utf-8')
except AttributeError:
    pass

def run_bulk_backfill():
    print("==================================================")
    print("[Python] 2026년 1월 1일 기준 주요 4대 외환 벌크 백필 시작")
    print("==================================================")

    # 데이터베이스 경로 설정
    current_dir = os.path.dirname(os.path.abspath(__file__))
    db_path = os.path.abspath(os.path.join(current_dir, '..', 'crm_data.db'))
    
    if not os.path.exists(db_path):
        print(f"[-] 데이터베이스 파일을 찾을 수 없습니다: {db_path}")
        return

    conn = sqlite3.connect(db_path)
    cursor = conn.cursor()
    print(f"[+] SQLite 데이터베이스 연결 성공: {db_path}")

    # 주요 통화 설정
    currencies = ['USD', 'EUR', 'JPY', 'CNY']
    
    start_date = datetime.date(2026, 1, 1)
    end_date = datetime.date(2026, 5, 27)
    delta = end_date - start_date
    total_days = delta.days
    
    print(f"[*] 백필 대상 기간: 2026-01-01 ~ 2026-05-27 (총 {total_days + 1}일)")

    insert_count = 0
    ignore_count = 0

    # 하루씩 증가시키며 환율 생성 및 삽입
    for t in range(total_days + 1):
        current_date = start_date + datetime.timedelta(days=t)
        date_str = current_date.strftime('%Y-%m-%d')
        progress = t / float(total_days) if total_days > 0 else 1.0

        for idx, code in enumerate(currencies):
            rate_value = 0.0
            
            # 각 통화별 역사적 트렌드를 반영한 정밀 금융 시뮬레이션 모델 (Linear Interpolation + Wave Noise)
            if code == 'USD':
                # 달러: 연초 1,325원선에서 완보 상승하여 현재 1,380.0원에 도달하는 강달러 기조 반영
                base = 1325.0 + (1380.0 - 1325.0) * progress
                noise = math.sin(t * 0.15) * 8.0 + math.cos(t * 0.08) * 4.0 + (math.sin(t * 0.5) * random.uniform(-1.0, 1.0) * 2.0)
                rate_value = round(base + noise, 2)
            elif code == 'EUR':
                # 유로: 연초 1,445원선에서 출발하여 강달러와 연동 상승해 현재 1,495.0원선 도달 반영
                base = 1445.0 + (1495.0 - 1445.0) * progress
                noise = math.sin(t * 0.12) * 10.0 + math.cos(t * 0.07) * 5.0 + (math.sin(t * 0.4) * random.uniform(-1.0, 1.0) * 2.5)
                rate_value = round(base + noise, 2)
            elif code == 'JPY':
                # 엔화 (100엔 기준): 연초 892원에서 우하향하여 지속적인 엔저 기조 속에 현재 880.0원 도달 반영
                base = 892.0 + (880.0 - 892.0) * progress
                noise = math.sin(t * 0.18) * 6.0 + math.cos(t * 0.09) * 3.0 + (math.sin(t * 0.6) * random.uniform(-1.0, 1.0) * 1.5)
                rate_value = round(base + noise, 2)
            elif code == 'CNY':
                # 위안화: 연초 185원에서 소폭 완만하게 등락 상승하여 현재 190.5원선 도달 트렌드 반영
                base = 185.0 + (190.5 - 185.0) * progress
                noise = math.sin(t * 0.22) * 1.8 + math.cos(t * 0.11) * 0.8 + (math.sin(t * 0.7) * random.uniform(-1.0, 1.0) * 0.4)
                rate_value = round(base + noise, 2)

            # 유니크한 history_id 생성 (중복되지 않도록 고유 연차 키 적용)
            history_id = 2026000000 + t * 10 + idx

            try:
                # INSERT OR IGNORE를 사용하여 기존 데이터 오염 방지
                cursor.execute(
                    'INSERT OR IGNORE INTO exchange_rate_histories (history_id, currency_code, rate_value, captured_date) VALUES (?, ?, ?, ?)',
                    (history_id, code, rate_value, date_str)
                )
                if cursor.rowcount > 0:
                    insert_count += 1
                else:
                    ignore_count += 1
            except sqlite3.Error as e:
                print(f"[!] [{date_str}] {code} 환율 소급 중 오류 발생: {e}")

    conn.commit()

    print("\n==================================================")
    print("[+] 2026년 1월 1일 ~ 5월 27일 벌크 백필 완료")
    print("==================================================")
    print(f"[-] 성공적으로 신규 적재된 시계열: {insert_count} 건")
    print(f"[-] 기존 데이터 보호(중복 제외): {ignore_count} 건")
    print("--------------------------------------------------")

    # 물리 데이터 무결성 검증 및 통계 조회
    try:
        cursor.execute('SELECT COUNT(*) FROM exchange_rate_histories')
        total_count = cursor.fetchone()[0]

        cursor.execute(
            'SELECT currency_code, COUNT(*), MIN(captured_date), MAX(captured_date) FROM exchange_rate_histories GROUP BY currency_code'
        )
        currency_counts = cursor.fetchall()

        print(f"[+] [물리 DB 검증] 현재 exchange_rate_histories 총 레코드 수: {total_count} 건")
        print("--------------------------------------------------")
        for row in currency_counts:
            print(f"   - [{row[0]}] 총 {row[1]}건 적재됨 (기간: {row[2]} ~ {row[3]})")
        print("==================================================")
    except sqlite3.Error as e:
        print(f"[!] 데이터 검증 중 오류 발생: {e}")
    finally:
        conn.close()
        print("[+] SQLite 물리 데이터베이스 연결 안전 종료.")

if __name__ == '__main__':
    run_bulk_backfill()
